packages/client/hmi-client/src/components/widgets/share-project/tera-share-project.vue

```diff
@@ -161,25 +162,64 @@ function selectNewPermissionForUser(permission: string, userId: string) {
 
 async function setPermissions() {
 	visible.value = false;
-	selectedUsers.value.forEach(async ({ id }) => {
+	await selectedUsers.value.forEach(async ({ id }) => {
 		const permission = newSelectedUserPermissions.get(id);
 		if (permission) {
 			const currentPermission = permissions.value?.permissionUsers.find((u) => u.id === id)
 				?.relationship;
 			if (permission === 'remove') {
 				if (currentPermission) {
-					await useProjects().removePermissions(props.project.id, id, currentPermission);
+					if (await useProjects().removePermissions(props.project.id, id, currentPermission)) {
+						removeUser(id);
+					} else {
+						addUser(id);
+					}
 				}
 			} else if (currentPermission) {
-				await useProjects().updatePermissions(props.project.id, id, currentPermission, permission);
-			} else {
-				await useProjects().setPermissions(props.project.id, id, permission);
+				if (
+					await useProjects().updatePermissions(props.project.id, id, currentPermission, permission)
+				) {
+					if (permission === 'reader') {
+						removeUser(id);
+					} else {
+						addUser(id);
+					}
+				}
+			} else if (await useProjects().setPermissions(props.project.id, id, permission)) {
+				if (permission === 'writer') {
+					addUser(id);
+				}
 			}
 		}
 	});
 	newSelectedUserPermissions.clear();
 }
 
+async function removeUser(id) {
+	const user = users.value.find((u) => u.id === id);
+	const name = `${user?.firstName} ${user?.lastName}`;
+	const updatedProject = cloneDeep(props.project);
+	if (updatedProject.authors) {
+		const index = updatedProject.authors.indexOf(name);
+		if (index !== undefined && index > -1) {
+			updatedProject.authors.splice(index, 1);
+		}
+		await useProjects().update(updatedProject);
+	}
+}
+
```

Comment:
I don't think that these remove/add user functions are technically doing anything.  The `Authors` field is set as read only/transient/optional and is only updated/set when we `GET` a project. Changing the authors list and updating won't actually change the object in the database.

Reply:
It is to update the front-end - which this currently forces the UI to refetch from the server, though maybe in a future there would be an "intelligent" update that does not do complete refetch.  Otherwise, the UI elements that are not the Share Dialog do not know that anything changed, and short of the user clicking the refresh page button, will not reflect the changes that the user just performed in said dialog.